### PR TITLE
update opencv dnn to support cann version >=6.3

### PR DESCRIPTION
**Pull request by SaltFish-T**
1.modify the search path of "libopsproto.so" in OpenCVFindCANN.cmake
2.add the search path of "libgraph_base.so" in OpenCVFindCANN.cmake
3.automatic check Ascend socVersion,and test on Ascend310/Ascend310B/Ascend910B well

### CODE REVIEW

File: cmake/OpenCVFindCANN.cmake [hunk 1]
```diff
@@ -58,7 +58,7 @@ if(CANN_INSTALL_DIR)
         return()
     endif()
     #  * libopsproto.so
-    set(lib_opsproto "${CANN_INSTALL_DIR}/opp/op_proto/built-in")
+    set(lib_opsproto "${CANN_INSTALL_DIR}/opp/built-in/op_proto/lib/linux/${CMAKE_HOST_SYSTEM_PROCESSOR}")
```
[the remark this answers is not in the document]
Comment (SaltFish-T) [reply cmake/OpenCVFindCANN.cmake:61]: ${CMAKE_HOST_SYSTEM_PROCESSOR} = x86_64 or aarch64. This depends on which CPU architecture is used

[the remark this answers is not in the document]
Comment (asmorkalov) [reply cmake/OpenCVFindCANN.cmake:61]: I mean, that different versions of CANN SDK has different layout. M.b. we need to presume compatibility and have both branches with version check?

[the remark this answers is not in the document]
Comment (SaltFish-T) [reply cmake/OpenCVFindCANN.cmake:61]: It's a good way. In my tests, this change is uitable for cann version 6.3 and above.

File: cmake/OpenCVFindCANN.cmake [hunk 2]
```diff
@@ -5,6 +5,13 @@ if("cann${CANN_INSTALL_DIR}" STREQUAL "cann" AND DEFINED ENV{ASCEND_TOOLKIT_HOME
     message(STATUS "CANN: updated CANN_INSTALL_DIR from ASCEND_TOOLKIT_HOME=$ENV{ASCEND_TOOLKIT_HOME}")
 endif()
 
+if(EXISTS "${CANN_INSTALL_DIR}/opp/op_proto/built-in/inc")
+    set(LOW_CANN_VERSION "YES" )
```
Comment (fengyuentau) [inline cmake/OpenCVFindCANN.cmake:9]: Can we use `CANN_VERSION_BELOW_6_3_ALPHA002` to make it meaningful?

File: cmake/OpenCVFindCANN.cmake [hunk 3]
```diff
@@ -58,24 +65,55 @@ if(CANN_INSTALL_DIR)
...
         message(STATUS "CANN: libopsproto.so is found at ${lib_opsproto}")
     else()
-        message(STATUS "CANN: Missing libopsproto.so. Turning off HAVE_CANN")
+        message(STATUS "CANN: Missing libopsproto.so can't found at ${lib_opsproto}. Turning off HAVE_CANN")
         set(HAVE_CANN OFF)
         return()
     endif()
-
+	  #  * lib_graph_base.so 
+	  set(lib_graph_base "${CANN_INSTALL_DIR}/compiler/lib64")
```
Comment (fengyuentau) [inline cmake/OpenCVFindCANN.cmake]: Does this library exist in version below 6.3.rc1.alpha002?

File: modules/dnn/src/net_cann.cpp [hunk 4]
```diff
@@ -304,9 +304,9 @@ std::shared_ptr<ge::ModelBufferData> compileCannGraph(std::shared_ptr<ge::Graph>
         bool ok;
         if ((child=fork()) == 0)
         {
-            // initialize engine
+            // initialize engine   Ascend310/Ascend310P3/Ascend910B/Ascend310B
             std::map<ge::AscendString, ge::AscendString> options = {
-                {ge::AscendString(ge::ir_option::SOC_VERSION), ge::AscendString("Ascend310")},
+                {ge::AscendString(ge::ir_option::SOC_VERSION), ge::AscendString(aclrtGetSocName())},
```
Comment (fengyuentau) [inline modules/dnn/src/net_cann.cpp:309]: Is `aclrtGetSocName()` available in all CANN versions?

Comment (fengyuentau) [inline modules/dnn/src/net_cann.cpp:309]: Is `aclrtGetSocName()` available in all CANN versions?

File: cmake/OpenCVFindCANN.cmake (hunk 2 above)
Comment (SaltFish-T) [reply cmake/OpenCVFindCANN.cmake:9]: OK,this makes it easier to understand

File: modules/dnn/src/net_cann.cpp (hunk 4 above)
Comment (SaltFish-T) [reply modules/dnn/src/net_cann.cpp:309]: Yes,it is available in all CANN versions.

**Review by fengyuentau — APPROVED** [review]
👍 Tried with this patch with old CANN and the build was successful.

**Review by fengyuentau — APPROVED** [review]
👍 Tried with this patch with old CANN and the build was successful.

File: cmake/OpenCVFindCANN.cmake [hunk 5]
```diff
@@ -58,24 +65,44 @@ if(CANN_INSTALL_DIR)
         return()
     endif()
     #  * libopsproto.so
-    set(lib_opsproto "${CANN_INSTALL_DIR}/opp/op_proto/built-in")
+    if (CANN_VERSION_BELOW_6_3_ALPHA002)
+        set(lib_opsproto "${CANN_INSTALL_DIR}/opp/op_proto/built-in/")
+    else()
+        if(EXISTS "${CANN_INSTALL_DIR}/opp/built-in/op_proto/lib/linux")
+            set(lib_opsproto "${CANN_INSTALL_DIR}/opp/built-in/op_proto/lib/linux/${CMAKE_HOST_SYSTEM_PROCESSOR}")
```
Comment (asmorkalov) [inline cmake/OpenCVFindCANN.cmake:72]: What about Windows? Forcing Linux dependency is bad idea. You can add `If (UNIX)` and `if (WIN32)` checks.

File: cmake/OpenCVFindCANN.cmake [hunk 6]
```diff
@@ -5,6 +5,13 @@ if("cann${CANN_INSTALL_DIR}" STREQUAL "cann" AND DEFINED ENV{ASCEND_TOOLKIT_HOME
     message(STATUS "CANN: updated CANN_INSTALL_DIR from ASCEND_TOOLKIT_HOME=$ENV{ASCEND_TOOLKIT_HOME}")
 endif()
 
+if(EXISTS "${CANN_INSTALL_DIR}/opp/op_proto/built-in/inc")
+    set(CANN_VERSION_BELOW_6_3_ALPHA002 "YES" )
+endif()
+if(CANN_VERSION_BELOW_6_3_ALPHA002)
+    add_definitions(-DCANN_VERSION_BELOW_6_3_ALPHA002="YES")
+endif()
```
Comment (asmorkalov) [inline cmake/OpenCVFindCANN.cmake:13]: Please merge the conditions.

File: cmake/OpenCVFindCANN.cmake (hunk 5 above)
Comment (SaltFish-T) [reply cmake/OpenCVFindCANN.cmake:70]: There is no cann version for Windows, and the "linux" is only folder under the path of  "${CANN_INSTALL_DIR}/opp/built-in/op_proto/lib/linux". So I wonder if it's necessary to make a check?

Comment (asmorkalov) [reply cmake/OpenCVFindCANN.cmake:70]: Please add `UNIX` check and warning at the very beginning.

File: cmake/OpenCVFindCANN.cmake [hunk 7]
```diff
@@ -5,7 +5,19 @@ if("cann${CANN_INSTALL_DIR}" STREQUAL "cann" AND DEFINED ENV{ASCEND_TOOLKIT_HOME
...
 
+if(EXISTS "${CANN_INSTALL_DIR}/opp/op_proto/built-in/inc")
+    set(CANN_VERSION_BELOW_6_3_ALPHA002 "YES" )
+    add_definitions(-DCANN_VERSION_BELOW_6_3_ALPHA002="YES")
+endif()
+
 if(CANN_INSTALL_DIR)
+    # Supported system: UNIX
+    if(UNIX)
+    else()
```
Comment (asmorkalov) [inline cmake/OpenCVFindCANN.cmake:16]: `if (NOT UNIX)`

File: cmake/OpenCVFindCANN.cmake [hunk 8]
```diff
@@ -5,7 +5,19 @@ if("cann${CANN_INSTALL_DIR}" STREQUAL "cann" AND DEFINED ENV{ASCEND_TOOLKIT_HOME
...
+        set(HAVE_CANN OFF)
+        message(STATUS "CANN: CANN toolkit supports unix but not ${CMAKE_SYSTEM_NAME}. Turning off HAVE_CANN")
```
Comment (asmorkalov) [inline cmake/OpenCVFindCANN.cmake:18]: I propose to use message(WARNING). It's highlighted red and caught by CI.

Comment (SaltFish-T) [reply cmake/OpenCVFindCANN.cmake]: I use message(WARNING) in every place that "set(HAVE_CANN OFF)".

**Review by asmorkalov — APPROVED** [review]
:+1:

**Pull request merged**